### ISSUE TITLE
How to setup pseudo-OOB Oauth properly?

### ISSUE THREAD

**Issue by ThoDuyNguyen** [r-lib/gargle#230]
Hi all 

We have a Ubuntu server which has RStudio Server installed. The server also has many cron jobs manipulating in Google Drive (using *googledrive*, *googlesheets*) and send email through Gmail (using *gmailr*) . 

I'm migrating from conventional OOB-auth to **pseudo-OOB auth** using these links below:

https://github.com/r-lib/gargle/blob/main/vignettes/auth-from-web.Rmd
https://github.com/r-lib/gargle/blob/main/vignettes/non-interactive-auth.Rmd

My purpose is to setup the cached authenticate for my crons could be executed non-interactively. The old setup (conventional OOB-oauth) works correctly up till now.  I use RStudio Server when running the code. 

### Setup the OAuth client

"https://www.tidyverse.org/google-callback" was used.

![image](https://user-images.githubusercontent.com/9877561/214478577-c72ad06d-8ef5-4d8c-bafd-53804f1be0b8.png)

![image](https://user-images.githubusercontent.com/9877561/214478436-108bec8d-4a95-4c4d-8259-a9903242786f.png)

### Initialized running in RStudio Server

```{r}
options(gargle_oauth_cache = "/path/to/.secret")
googledrive::drive_auth(path = "/path/to/downloaded/json/file")
```
![image](https://user-images.githubusercontent.com/9877561/214479281-eab326f0-fc7a-44ed-a96f-28b3494340b8.png)

![image](https://user-images.githubusercontent.com/9877561/214480115-3d966f22-b491-4876-9199-26166c2ad6aa.png)

```
trying `token_fetch()`
trying `credentials_service_account()`
JSON does not appear to represent a service account
Did you provide the JSON for an OAuth client instead of for a service account?
trying `credentials_external_account()`
aws.ec2metadata not installed; can't detect whether running on EC2 instance
trying `credentials_app_default()`
trying `credentials_gce()`
trying `credentials_byo_oauth()`
Error caught by `token_fetch()`:
inherits(token, "Token2.0") is not TRUE
trying `credentials_user_oauth2()`
Gargle2.0 initialize
attempt to access internal gargle data from: googledrive
adding "userinfo.email" scope
loading token from the cache
no matching token in the cache
initiating new token
Enter authorization code: <remove>
putting token into the cache:
</path/to/.secret>
```

After that, R scripts could run by `Rscript --vanilla` 

```{r}
options(
  gargle_verbosity = "debug",
  gargle_oauth_cache = "/path/to/.secret"
)

URL <- "REMOVE"

df_file <- googledrive::drive_ls(URL)

lapply(df_file$id[1:2], function(file_id) {
  cli::cli_alert_info(as.character(file_id))
  googledrive::drive_download(file_id, overwrite = TRUE, verbose = FALSE)
  return(TRUE)
})

``` 

Output 

```
initializing AuthState
attempt to access internal gargle data from: googledrive
trying `token_fetch()`
trying `credentials_service_account()`
Error caught by `token_fetch()`:
Argument 'txt' must be a JSON string, URL or file.
trying `credentials_external_account()`
aws.ec2metadata not installed; can't detect whether running on EC2 instance
trying `credentials_app_default()`
trying `credentials_gce()`
trying `credentials_byo_oauth()`
Error caught by `token_fetch()`:
inherits(token, "Token2.0") is not TRUE
trying `credentials_user_oauth2()`
Gargle2.0 initialize
attempt to access internal gargle data from: googledrive
adding "userinfo.email" scope
loading token from the cache
! Using an auto-discovered, cached token.
  To suppress this message, modify your code or options to clearly consent to the use of a cached token.
  See gargle's "Non-interactive auth" vignette for more details:
  <https://gargle.r-lib.org/articles/non-interactive-auth.html>
ℹ The googledrive package is using a cached token for <REMOVE>
matching token found in the cache
attempt to access internal gargle data from: googledrive
ℹ <REMOVE>
attempt to access internal gargle data from: googledrive
attempt to access internal gargle data from: googledrive
ℹ <REMOVE>
attempt to access internal gargle data from: googledrive
attempt to access internal gargle data from: googledrive
[[1]]
[1] TRUE

[[2]]
[1] TRUE

Warning message:
The `verbose` argument of `some_googledrive_function()` is deprecated as of googledrive 2.0.0.
ℹ Set `options(googledrive_quiet = TRUE)` to suppress all googledrive messages. For finer control, use `local_drive_quiet()` or `with_drive_quiet()`. googledrive's `verbose` argument will be removed in the future.
ℹ The deprecated feature was likely used in the googledrive package.
  Please report the issue at <https://github.com/tidyverse/googledrive/issues>. 
```

My session info
```
R version 4.2.2 Patched (2022-11-10 r83330)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.1 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8      
 [8] LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] fansi_1.0.3       assertthat_0.2.1  dplyr_1.0.10      utf8_1.2.2        R6_2.5.1          DBI_1.1.3         lifecycle_1.0.3   magrittr_2.0.3    pillar_1.8.1      rlang_1.0.6       cli_3.4.1        
[12] rstudioapi_0.14   fs_1.5.2          generics_0.1.3    vctrs_0.4.2       tools_4.2.2       googledrive_2.0.0 glue_1.6.2        purrr_0.3.5       compiler_4.2.2    pkgconfig_2.0.3   gargle_1.2.1     
[23] tidyselect_1.1.2  tibble_3.1.8    
```
 
The flow and output seem similar to conventional OOB-OAuth (deprecated soon). Would someone help me confirm this is the proper way to change to  pseudo-OOB OAuth?

Thank you in advance

**Comment by jennybc** [comment]
Working through what I see above in order, this is a mistake:

``` r
googledrive::drive_auth(path = "/path/to/downloaded/json/file")
```

Any file provided directly to `drive_auth()` must be a service account token.

https://googledrive.tidyverse.org/reference/drive_auth.html

> `path` JSON identifying the service account, in one of the forms supported for the txt argument of [jsonlite::fromJSON()](https://rdrr.io/pkg/jsonlite/man/fromJSON.html) (typically, a file path or JSON string).

But you want to configure an OAuth application, which you will then use to obtain (and cache) a token.

That is what `drive_auth_configure()` is for.

https://googledrive.tidyverse.org/reference/drive_auth_configure.html

> `path` JSON downloaded from Google Cloud Platform Console, containing a client id (aka key) and secret, in one of the forms supported for the txt argument of [jsonlite::fromJSON()](https://rdrr.io/pkg/jsonlite/man/fromJSON.html) (typically, a file path or JSON string).

I can't digest the rest right now, but this alone may get you unstuck.

**Comment by ThoDuyNguyen** [comment]
Thank you for the reply. 

I try another attempt but it did not work. 

I use the code from 

https://github.com/r-lib/gargle/pull/213
https://github.com/r-lib/gargle/pull/214

```
options(
  gargle_oauth_cache = "/home/thond/gargle_json_cache/drive/.secrets",
  gargle_verbosity = "debug",
  gargle_oob_default = TRUE
)

the_app <- gargle::oauth_app_from_json("/path/to/json/download/from/google/console")

## Checking the_app's content

googledrive::drive_auth_configure(the_app)

googledrive::drive_oauth_app()

googledrive::drive_auth(email = NA, use_oob = TRUE)

```

I think there is something weird in **redirect_uris**

Output comsole
```

trying `token_fetch()`
trying `credentials_service_account()`
Error caught by `token_fetch()`:
Argument 'txt' must be a JSON string, URL or file.
trying `credentials_external_account()`
aws.ec2metadata not installed; can't detect whether running on EC2 instance
trying `credentials_app_default()`
trying `credentials_gce()`
trying `credentials_byo_oauth()`
Error caught by `token_fetch()`:
inherits(token, "Token2.0") is not TRUE
trying `credentials_user_oauth2()`
Gargle2.0 initialize
adding "userinfo.email" scope
loading token from the cache
no matching token in the cache
initiating new token
initializing AuthState
initializing AuthState
Enter authorization code: 
```

Output 
![image](https://user-images.githubusercontent.com/9877561/214752951-a940bdcc-518d-4450-8186-8b0dc8203933.png)

Error detail
![image](https://user-images.githubusercontent.com/9877561/214753237-a4d3816b-f8d9-438d-88dc-076e2f9260fd.png)

Setting of OAuth client 
![image](https://user-images.githubusercontent.com/9877561/214753579-b151585c-599d-4119-b8e1-f8420b6daa5f.png)

Downloaded json:
![image](https://user-images.githubusercontent.com/9877561/214753634-1ed66366-1f9d-4177-a719-307353098c9c.png)

I look into the code and found out that: 

```
str(the_app)
```
Output
```
List of 4
 $ appname     : chr <identical with project_id in JSON>
 $ secret      : chr <identical with client_secret in JSON>
 $ key         : chr <identical with client_id in JSON>
 $ redirect_uri: chr "http://localhost:1410/"     => incorrect      
 - attr(*, "class")= chr "oauth_app"
```

I trace the code a little bit 
Code
```
gargle::oauth_app_from_json
```
Output
```
function (path, appname = NULL) 
{
    stopifnot(is_string(path), is.null(appname) || is_string(appname))
    json <- jsonlite::fromJSON(path, simplifyVector = FALSE)
    info <- json[["installed"]] %||% json[["web"]]
    if (!all(c("client_id", "client_secret") %in% names(info))) {
        gargle_abort("\n      Can't find {.field client_id} and {.field client_secret} in the JSON.")
    }
    httr::oauth_app(appname = appname %||% info$project_id, key = info$client_id, 
        secret = info$client_secret)
}
<bytecode: 0x557ef2a143f8>
<environment: namespace:gargle>
```
Code
```
httr::oauth_app
```
Ouput
```
function (appname, key, secret = NULL, redirect_uri = oauth_callback()) 
{
    if (missing(secret)) {
        env_name <- paste0(toupper(appname), "_CONSUMER_SECRET")
        secret <- Sys.getenv(env_name)
        if (secret == "") {
            warning("Couldn't find secret in environment variable ", 
                env_name, call. = FALSE)
            secret <- NULL
        }
        else {
            message("Using secret stored in environment variable ", 
                env_name)
        }
    }
    structure(list(appname = appname, secret = secret, key = key, 
        redirect_uri = redirect_uri), class = "oauth_app")
}
<bytecode: 0x557ef2f6be70>
<environment: namespace:httr>
```
Code
```
httr::oauth_callback()
```
Output
```
"http://localhost:1410/"  
```
Since the error message **"The redirect URI in the request, urn:ietf:wg:oauth:2.0:oob, can only be used by a Client ID for native application. It is not allowed for the WEB client type"** so I think the problem could be redirect_uri did not get from the JSON file.  I try to force it 
```
the_app[["redirect_uri"]] <- "https://www.tidyverse.org/google-callback"
```

Follow the code in 
https://github.com/r-lib/gargle/blob/52938d6ccdba9f6436f9fe520d4dce1850542531/R/oauth-init.R

```
  if (use_oob) {
    redirect_uri <- oob_value %||% "urn:ietf:wg:oauth:2.0:oob"

    # https://developers.googleblog.com/2022/02/making-oauth-flows-safer.html#warning-message
    # "The developers can acknowledge the user-facing warning message and
    # suppress it by passing a query parameter in the authorization call ..."
    client_name <- app$name %||% app$appname %||% ""
    if (client_name %in% c("tidyverse-clio", "gargle-clio")) {
      query_authorize_extra[["ack_oob_shutdown"]] <- "2022-10-03"
    }
```

In https://github.com/r-lib/gargle/blob/52938d6ccdba9f6436f9fe520d4dce1850542531/R/Gargle-class.R
```
  if (use_oob && identical(client_type, "web")) {
    params$oob_value <- select_pseudo_oob_value(app$redirect_uris)
  }
```

the_app use *redirect_uri* not **redirect_uris**. Therefore **oob_value** could be NULL => use default redirect_uri.

**Comment by ThoDuyNguyen** [comment]
I use the development version and it work perfectly. It seems the CRAN version still not support the current flow (My bad)

My working code is: 

```
options(
  gargle_oauth_cache = "/path/to/cache",
  gargle_verbosity = "debug",
  gargle_oob_default = TRUE,
  gargle_oauth_email = "email" 
)

the_app <- gargle::gargle_oauth_client_from_json("/path/to/download/json")

googledrive::drive_auth_configure(the_app)

googledrive::drive_auth(use_oob = TRUE)

URL <- "<URL>"

df_file <- googledrive::drive_ls(URL)

lapply(df_file$id, function(file_id) {
  googledrive::drive_download(file_id, overwrite = TRUE)
  return(TRUE)
})
```
In case I run the code non-interactive (usinmg Rscript), when the cache is expired, the code auto refresh the cache automatically. The document does not work for me: 

![image](https://user-images.githubusercontent.com/9877561/214771414-cfa8073f-0b65-483d-8ac4-459a31fec299.png)
 
I must include these lines in the Script. 

```
the_app <- gargle::gargle_oauth_client_from_json("/path/to/download/json")

googledrive::drive_auth_configure(the_app)

googledrive::drive_auth(use_oob = TRUE)
```

Thank you in advance

**Comment by jennybc** [comment]
> I use the development version and it work perfectly. It seems the CRAN version still not support the new flow.

Yes, true! I am releasing gargle in the next couple days.

**Comment by jennybc** [comment]
I think this thread is resolved by releasing gargle 1.3.0.
